R/ZIPLNfit-class.R

```diff
@@ -0,0 +1,831 @@
+
+      ## Initialization of the PLN component
+      private$B <- B
+      private$M <- M
+      private$S <- matrix(.1, n, p)
+
+      ## Link to functions performing the optimization
+      private$optimizer$B     <- function(M, X, Omega, control) optim_zipln_B_dense(M, X)
+      private$optimizer$zi    <- switch(
+        control$ziparam,
+        "single" = function(init_B0, X0, R, config) list(Pi = matrix(    mean(R), nrow(R), p)              , B0 = matrix(NA, d0, p)),
+        "row"    = function(init_B0, X0, R, config) list(Pi = matrix(rowMeans(R), nrow(R), p)              , B0 = matrix(NA, d0, p)),
+        "col"    = function(init_B0, X0, R, config) list(Pi = matrix(colMeans(R), nrow(R), p, byrow = TRUE), B0 = matrix(NA, d0, p)),
+        "covar"  = optim_zipln_zipar_covar
+      )
+      private$optimizer$Omega <- optim_zipln_Omega_full
+
+    },
+
+    #' @description Call to the Cpp optimizer and update of the relevant fields
+    #' @param control a list for controlling the optimization. See details.
+    optimize = function(responses, covariates, offsets, weights, control) {
+
+      data <- list(Y = responses, X = covariates$PLN, X0 = covariates$ZI, O = offsets)
+      parameters <-
+        list(Omega = NA, B0 = private$B0, B = private$B, Pi = private$Pi,
+             M = private$M, S = private$S, R = private$R)
+
+      # Main loop
+      nb_iter <- 0
+      criterion   <- numeric(control$maxit_out)
+      convergence <- numeric(control$maxit_out)
+
+      vloglik <- -Inf; objective <- Inf
+      repeat {
+
+        # Check maxeval
+        if (control$maxit_out >= 0 && nb_iter >= control$maxit_out) {
+          stop_reason <- "maximum number of iterations reached"
+          criterion   <- criterion[1:nb_iter]
+          convergence <- convergence[nb_iter]
+          break
+        }
+
+        # M Step
+        new_Omega <- private$optimizer$Omega(
+          M = parameters$M, X = data$X, B = parameters$B, S = parameters$S
+        )
+        new_B <- private$optimizer$B(
+          M = parameters$M, X = data$X, Omega = new_Omega, control
+        )
+
+        optim_new_zipar <- private$optimizer$zi(
+          init_B0 = parameters$B0, X0 = data$X0, R = parameters$R, config = control
+        )
+        new_B0 <- optim_new_zipar$B0
+        new_Pi <- optim_new_zipar$Pi
+
+        # VE Step
+        new_R <- optim_zipln_R(
+          Y = data$Y, X = data$X, O = data$O, M = parameters$M, S = parameters$S, Pi = new_Pi
+        )
+        optim_new_M <- optim_zipln_M(
+          init_M = parameters$M,
+          Y = data$Y, X = data$X, O = data$O, R = new_R, S = parameters$S, B = new_B, Omega = new_Omega,
+          configuration = control
+        )
+        new_M <- optim_new_M$M
+        optim_new_S <- optim_zipln_S(
+          init_S = parameters$S,
+          O = data$O, M = new_M, R = new_R, B = new_B, diag_Omega = diag(new_Omega),
+          configuration = control
+        )
+        new_S <- optim_new_S$S
```

Comment:
Why not do directly `new_S <- optim_zipln_S(...)$S` instead of using the intermediate `optim_new_S` ?

Reply:
See above, same answer.